P S N Madhav 17BCD7004

In [40]:
import pandas as pd
import numpy as np

In [41]:
data=pd.read_csv('Rainfall_final.csv')
data.head()

year  month  day  temp high  tempavg  temp low  DP high  DPavg  DP low  \
0  2008      1    1         24       21        18       17     15      12   
1  2008      1    2         26       22        19       18     17      16   
2  2008      1    3         25       21        18       16     16      13   
3  2008      1    4         26       23        20       17     16      15   
4  2008      1    5         26       21        17       17     15      14   

   humidity high    ...     humidity low  SLP high  SLPavg  SLP low  \
0             83    ...               47      1015    1013     1011   
1             88    ...               54      1016    1014     1012   
2             88    ...               47      1017    1014     1006   
3             83    ...               54      1016    1014     1013   
4             88    ...               47      1017    1015     1012   

   visibility high  visibilityavg  visibilitylow  wind high  windavg  Rainfall  
0                5              4              1         19        8         0  
1                5              4              1         11        2         0  
2                5              4              0         19       10         1  
3                5              5              4         14       11         0  
4                5              4              1         23       11         0  

[5 rows x 21 columns]

In [42]:
data.isnull().sum()

year               0
month              0
day                0
temp high          0
tempavg            0
temp low           0
DP high            0
DPavg              0
DP low             0
humidity high      0
humidity avg       0
humidity low       0
SLP high           0
SLPavg             0
SLP low            0
visibility high    0
visibilityavg      0
visibilitylow      0
wind high          0
windavg            0
Rainfall           0
dtype: int64

In [43]:
data.shape

(3604, 21)

In [44]:
x=data.drop(['Rainfall'], axis=1)
y=data.Rainfall.values

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size= 0.25, random_state = 0)

In [45]:
# Logistic Regression
from sklearn.linear_model import LogisticRegression
from sklearn import metrics 
logistic_regression = LogisticRegression()
logistic_regression.fit(x_train,y_train)
y_predlr = logistic_regression.predict(x_test)

from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
print("confusion matrix:-")
print(confusion_matrix(y_test, y_predlr))
print("accuracy score:-",accuracy_score(y_test, y_predlr))

print("classification report:-")
print(classification_report(y_test, y_predlr))

confusion matrix:-
[[449  83]
 [108 261]]
accuracy score:- 0.7880133185349611
classification report:-
             precision    recall  f1-score   support

          0       0.81      0.84      0.82       532
          1       0.76      0.71      0.73       369

avg / total       0.79      0.79      0.79       901



In [46]:
# SVM

from sklearn import svm
clf = svm.SVC(kernel='linear') # Linear Kernel

#Train the model using the training sets
clf.fit(x_train, y_train)

y_predsvc = clf.predict(x_test)
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
print("confusion matrix:-")
print(confusion_matrix(y_test, y_predsvc))
print("accuracy score:-",accuracy_score(y_test, y_predsvc))

print("classification report:-")
print(classification_report(y_test, y_predsvc))

confusion matrix:-
[[460  72]
 [113 256]]
accuracy score:- 0.7946725860155383
classification report:-
             precision    recall  f1-score   support

          0       0.80      0.86      0.83       532
          1       0.78      0.69      0.73       369

avg / total       0.79      0.79      0.79       901



In [47]:
# SVM with rbf

from sklearn.svm import SVC
rbfclassifier = SVC(kernel='rbf')
rbfclassifier.fit(x_train, y_train)



y_predksvc = rbfclassifier.predict(x_test)
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
print("confusion matrix:-")
print(confusion_matrix(y_test, y_predksvc))
print("accuracy score:-",accuracy_score(y_test, y_predksvc))

print("classification report:-")
print(classification_report(y_test, y_predksvc))

confusion matrix:-
[[512  20]
 [307  62]]
accuracy score:- 0.6370699223085461
classification report:-
             precision    recall  f1-score   support

          0       0.63      0.96      0.76       532
          1       0.76      0.17      0.27       369

avg / total       0.68      0.64      0.56       901



In [48]:
# SVM with sigmoid

from sklearn.svm import SVC
rbfclassifier = SVC(kernel='sigmoid')
rbfclassifier.fit(x_train, y_train)



y_predksvc = rbfclassifier.predict(x_test)
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
print("confusion matrix:-")
print(confusion_matrix(y_test, y_predksvc))
print("accuracy score:-",accuracy_score(y_test, y_predksvc))

print("classification report:-")
print(classification_report(y_test, y_predksvc))

confusion matrix:-
[[532   0]
 [369   0]]
accuracy score:- 0.5904550499445061
classification report:-
             precision    recall  f1-score   support

          0       0.59      1.00      0.74       532
          1       0.00      0.00      0.00       369

avg / total       0.35      0.59      0.44       901



C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [49]:
# Decision Tree

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn import tree

# Create Decision Tree classifer object
clf = DecisionTreeClassifier()

# Train Decision Tree Classifer
clf = clf.fit(x_train,y_train)

#Predict the response for test dataset
y_pre = clf.predict(x_test)

from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
print("confusion matrix:-")
print(confusion_matrix(y_test, y_pre))
print("accuracy score:-",accuracy_score(y_test, y_pre))

print("classification report:-")
print(classification_report(y_test, y_pre))

confusion matrix:-
[[410 122]
 [ 95 274]]
accuracy score:- 0.7591564927857936
classification report:-
             precision    recall  f1-score   support

          0       0.81      0.77      0.79       532
          1       0.69      0.74      0.72       369

avg / total       0.76      0.76      0.76       901



In [50]:
# KNN

#Import knearest neighbors Classifier model
from sklearn.neighbors import KNeighborsClassifier

#Create KNN Classifier
knn = KNeighborsClassifier(n_neighbors=5)

#Train the model using the training sets
knn.fit(x_train, y_train)

#Predict the response for test dataset
y_predknn = knn.predict(x_test)

from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
print("confusion matrix:-")
print(confusion_matrix(y_test, y_predknn))
print("accuracy score:-",accuracy_score(y_test, y_predknn))

print("classification report:-")
print(classification_report(y_test, y_predknn))

confusion matrix:-
[[423 109]
 [127 242]]
accuracy score:- 0.7380688124306326
classification report:-
             precision    recall  f1-score   support

          0       0.77      0.80      0.78       532
          1       0.69      0.66      0.67       369

avg / total       0.74      0.74      0.74       901



Accuracy of different classifications :

Logistic_Regression = 0.7880133185349611
SVM with linear     = 0.7946725860155383
SVM with rbf        = 0.6370699223085461
SVM with sigmoid    = 0.5904550499445061
Decision Tree       = 0.7502774694783574
KNN                 = 0.7380688124306326